ssl/quic/quic_impl.c

```diff
@@ -9,271 +9,1244 @@
+        tv->tv_usec = 0;
+        return 1;
+    }
+
+    now = ossl_time_now();
+    if (ossl_time_compare(now, deadline) >= 0) {
+        tv->tv_sec  = 0;
+        tv->tv_usec = 0;
+        return 1;
+    }
+
+    *tv = ossl_time_to_timeval(ossl_time_subtract(deadline, now));
+    return 1;
+}
 
-    if (sc == NULL || rbio == NULL)
+/* SSL_get_rpoll_descriptor */
+int ossl_quic_get_rpoll_descriptor(QUIC_CONNECTION *qc, BIO_POLL_DESCRIPTOR *desc)
+{
+    if (desc == NULL || qc->net_rbio == NULL)
         return 0;
 
-    sc->rwstate = SSL_READING;
-    ret = BIO_read_ex(rbio, buf, len, readbytes);
-    if (ret > 0 || !BIO_should_retry(rbio))
-        sc->rwstate = SSL_NOTHING;
-    return ret <= 0 ? -1 : ret;
+    return BIO_get_rpoll_descriptor(qc->net_rbio, desc);
 }
 
-int ossl_quic_peek(SSL *s, void *buf, size_t len, size_t *readbytes)
+/* SSL_get_wpoll_descriptor */
+int ossl_quic_get_wpoll_descriptor(QUIC_CONNECTION *qc, BIO_POLL_DESCRIPTOR *desc)
 {
-    return -1;
+    if (desc == NULL || qc->net_wbio == NULL)
+        return 0;
+
+    return BIO_get_wpoll_descriptor(qc->net_wbio, desc);
 }
 
-int ossl_quic_write(SSL *s, const void *buf, size_t len, size_t *written)
+/* SSL_want_net_read */
+int ossl_quic_get_want_net_read(QUIC_CONNECTION *qc)
 {
-    BIO *wbio = SSL_get_wbio(s);
-    int ret;
-    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_QUIC_SSL(s);
+    if (qc->ch == NULL)
+        return 0;
 
-    if (sc == NULL || wbio == NULL)
+    return ossl_quic_reactor_want_net_read(ossl_quic_channel_get_reactor(qc->ch));
+}
+
+/* SSL_want_net_write */
+int ossl_quic_get_want_net_write(QUIC_CONNECTION *qc)
+{
+    if (qc->ch == NULL)
         return 0;
 
-    sc->rwstate = SSL_WRITING;
-    ret = BIO_write_ex(wbio, buf, len, written);
-    if (ret > 0 || !BIO_should_retry(wbio))
-        sc->rwstate = SSL_NOTHING;
-    return ret;
+    return ossl_quic_reactor_want_net_write(ossl_quic_channel_get_reactor(qc->ch));
 }
 
+/*
+ * QUIC Front-End I/O API: Connection Lifecycle Operations
+ * =======================================================
+ *
+ *         SSL_do_handshake         => ossl_quic_do_handshake
+ *         SSL_set_connect_state    => ossl_quic_set_connect_state
+ *         SSL_set_accept_state     => ossl_quic_set_accept_state
+ *         SSL_shutdown             => ossl_quic_shutdown
+ *         SSL_ctrl                 => ossl_quic_ctrl
+ *   (BIO/)SSL_connect              => ossl_quic_connect
+ *   (BIO/)SSL_accept               => ossl_quic_accept
+ *
+ */
+
+/* SSL_shutdown */
 int ossl_quic_shutdown(SSL *s)
 {
+    QUIC_CONNECTION *qc = QUIC_CONNECTION_FROM_SSL(s);
+
+    if (!expect_quic_conn(qc))
+        return 0;
+
+    if (qc->ch != NULL)
+        ossl_quic_channel_local_close(qc->ch);
+
     return 1;
 }
 
+/* SSL_ctrl */
+static void fixup_mode_change(QUIC_CONNECTION *qc)
+{
+    /* If enabling EPW mode, cancel any AON write */
+    if ((qc->ssl_mode & SSL_MODE_ENABLE_PARTIAL_WRITE) != 0)
+        aon_write_finish(qc);
```

Comment:
It seems like an error to me to enable partial write if an AON is half way done.

Reply:
We don't have any way to indicate error on mode change. What do you want to do?

Reply:
Rather than fail when we set the mode we could just fail here if `aon_write_in_progress != 0` - and document `SSL_MODE_ENABLE_PARTIAL_WRITE` to say you can't enable it if there is a write ongoing. This seems reasonable to say even for TLS.